File: CPP/42_TrappingRainWater.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        vector<int> lmx;
        vector<int> rmx;
        lmx.push_back(height[0]);
        for(int i=1;i<height.size();i++) 
            {
            lmx.push_back(max(lmx[i-1],height[i])) ;
        }
        rmx.push_back(height[height.size()-1]);
        int t=1;
        for(int i=height.size()-2;i>=0;i--) 
            {
            rmx.push_back(max(rmx[t-1],height[i]));
            t++;
        }
        reverse(rmx.begin(),rmx.end());
        int ans=0;
        for(int i=0;i<height.size();i++) 
            {
            ans+=min(lmx[i],rmx[i])-height[i];
        }
        return ans;
    }
};

File: Python/42_TrappingRainWater.py
class Solution:
    def trap(self, height: List[int]) -> int:
        lmx=[]
        rmx=[]
        lmx.append(height[0]) 
        rmx.append(height[-1]) 
        for i in range(1,len(height)):
            lmx.append(max(lmx[i-1],height[i])) 
        for i in range(len(height)-2,-1,-1):
                       rmx.insert(0,max(rmx[0],height[i])) 
        res=[]
        for i in range(len(height)):
                       res.append(min(lmx[i],rmx[i])-height [i]) 
        print(res) 
        return sum(res) 
       
        
